product-accelerators/apim/components/org.wso2.healthcare.apim.backendauth/src/main/java/org/wso2/healthcare/apim/backendauth/BackendAuthenticator.java

```diff
@@ -0,0 +1,103 @@
+ */
+
+package org.wso2.healthcare.apim.backendauth;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.apache.synapse.MessageContext;
+import org.apache.synapse.core.axis2.Axis2MessageContext;
+import org.apache.synapse.mediators.AbstractMediator;
+import org.wso2.healthcare.apim.core.OpenHealthcareEnvironment;
+import org.wso2.healthcare.apim.core.OpenHealthcareException;
+import org.wso2.healthcare.apim.core.config.BackendAuthConfig;
+
+import java.util.Map;
+
+/**
+ * Backend authenticator mediator.
+ */
+public class BackendAuthenticator extends AbstractMediator {
+
+    private static final Log log = LogFactory.getLog(BackendAuthenticator.class);
+    private String configName;
+    private final Map<String, BackendAuthConfig> backendAuthConfig;
+    private final PrivateKeyJWTBackendAuthenticator privateKeyJWTBackendAuthenticator;
+
+    public BackendAuthenticator() throws OpenHealthcareException {
+        backendAuthConfig = OpenHealthcareEnvironment.getInstance().getConfig().getBackendAuthConfig();
+        privateKeyJWTBackendAuthenticator = new PrivateKeyJWTBackendAuthenticator();
+
+    }
+
+    @Override
+    public boolean mediate(MessageContext messageContext) {
+        if (log.isDebugEnabled()) {
+            log.debug("Backend authenticator mediator is invoked.");
+        }
+
+        BackendAuthConfig config = backendAuthConfig.get(configName);
+        String accessToken;
+
+        if (configName == null) {
+            log.error("Auth type is not defined in the message context.");
+            return false;
+        }
+        if (config.getAuthType().equals(Constants.POLICY_ATTR_AUTH_TYPE_PKJWT)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Auth type is PKJWT.");
+            }
+            accessToken = privateKeyJWTBackendAuthenticator.fetchValidAccessToken(messageContext, config);
+        } else if (config.getAuthType().equals(Constants.POLICY_ATTR_AUTH_TYPE_CLIENT_CRED)) {
+            if (log.isDebugEnabled()) {
+                log.debug("Auth type is CLIENT CREDENTIALS.");
+            }
+            return true;
+        } else {
+            log.error("Auth type is not supported.");
+            return false;
+        }
+
+        if (messageContext instanceof Axis2MessageContext) {
+            org.apache.axis2.context.MessageContext axisMsgCtx =
+                    ((Axis2MessageContext) messageContext).getAxis2MessageContext();
+            Object headers = axisMsgCtx.getProperty(org.apache.axis2.context.MessageContext.TRANSPORT_HEADERS);
+            if (headers instanceof Map) {
+                Map headersMap = (Map) headers;
+                if (headersMap.get(Constants.HEADER_NAME_AUTHORIZATION) != null) {
+                    headersMap.remove(Constants.HEADER_NAME_AUTHORIZATION);
+                }
+                headersMap.put(Constants.HEADER_NAME_AUTHORIZATION, Constants.HEADER_VALUE_BEARER + accessToken);
```

Comment:
Let's get the authentication scheme(i.e 'Basic', 'Bearer') from the authenticator itself. This way someone can write a backend basic authenticator as well.

Reply:
changed in 
2baf7e5